influxdb_store.go

```diff
@@ -0,0 +1,409 @@
+		}
+		span.Annotations = *annotations
+		if span.ID.IsRoot() && rootSpanSet {
+			return nil, errors.New("unexpected multiple root spans")
+		}
+		if span.ID.IsRoot() && !rootSpanSet {
+			isRootSpan = true
+		}
+		if isRootSpan { // root span.
+			trace.Span = *span
+			rootSpanSet = true
+		} else { // children span.
+			trace.Sub = append(trace.Sub, &Trace{Span: *span})
+		}
+	}
+	return trace, nil
+}
+
+func (in *InfluxDBStore) Traces() ([]*Trace, error) {
+	traces := make([]*Trace, 0)
+
+	// GROUP BY * -> meaning group by all tags(trace_id, span_id & parent_id)
+	// grouping by all tags includes those and it's values on the query response.
+	rootSpansQuery := fmt.Sprintf("SELECT * FROM spans WHERE parent_id='%s' GROUP BY * LIMIT %d", zeroID, in.tracesPerPage)
+	rootSpansResult, err := in.executeOneQuery(rootSpansQuery)
+	if err != nil {
+		return nil, err
+	}
+
+	// result.Series -> A slice containing all the spans.
+	if len(rootSpansResult.Series) == 0 {
+		return traces, nil
+	}
+
+	// Cache to keep track of traces to be returned.
+	tracesCache := make(map[ID]*Trace, 0)
+
+	// Iterate over series(spans) to create traces.
+	for _, s := range rootSpansResult.Series {
+		span, err := newSpanFromRow(&s)
+		if err != nil {
+			return nil, err
+		}
+		annotations, err := annotationsFromRow(&s)
+		if err != nil {
+			return nil, err
+		}
+		span.Annotations = *annotations
+		_, present := tracesCache[span.ID.Trace]
+		if !present {
+			tracesCache[span.ID.Trace] = &Trace{Span: *span}
+		} else {
+			return nil, errors.New("duplicated root span")
+		}
+	}
+
+	// Using 'OR' since 'IN' not supported yet.
+	where := `WHERE `
+	var i int = 1
+	for _, trace := range tracesCache {
+		where += fmt.Sprintf("(trace_id='%s' AND parent_id!='%s')", trace.Span.ID.Trace, zeroID)
+
+		// Adds 'OR' except for last iteration.
+		if i != len(tracesCache) && len(tracesCache) > 1 {
+			where += " OR "
+		}
+		i += 1
+	}
+
+	// Queries for all children spans of the traces to be returned.
+	childrenSpansQuery := fmt.Sprintf("SELECT * FROM spans %s GROUP BY *", where)
+	childrenSpansResult, err := in.executeOneQuery(childrenSpansQuery)
+	if err != nil {
+		return nil, err
+	}
+
+	// Iterate over series(children spans) to create sub-traces
+	// and associates sub-traces with it's parent trace.
+	for _, s := range childrenSpansResult.Series {
+		span, err := newSpanFromRow(&s)
+		if err != nil {
+			return nil, err
+		}
+		annotations, err := annotationsFromRow(&s)
+		if err != nil {
+			return nil, err
+		}
+		span.Annotations = *annotations
+		trace, present := tracesCache[span.ID.Trace]
+		if !present { // Root trace not added.
+			return nil, errors.New("parent not found")
+		} else { // Root trace already added so append a sub-trace.
+			trace.Sub = append(trace.Sub, &Trace{Span: *span})
+		}
+	}
+	for _, trace := range tracesCache {
+		traces = append(traces, trace)
+	}
+	return traces, nil
+}
+
+func (in *InfluxDBStore) Close() {
+	in.server.Close()
```

Comment:
good call! - yes `in.server.Close()` does return an `error` so fixes on ab57107457f411147d7d1cddcf14d73f5fefa45f & b4a8ae4fab67e71c0a56371912ce08ca4442609f